### PR TITLE
Fix %p build placeholder, and various placeholder replacement fixes and improvements

### PR DESCRIPTION
**Pull request by b4n**
* Fix `%p` replacement in build (and print external) commands when no project is open (#4314)
* Fix replacing placeholders in replacements.  This happened for a long time, but in could cause fairly annoying issues if the replacement did contain a placeholder in an unexpected fashion, and that's usually unexpected with paths and alike.
* Support `%%` to insert a literal `%` in build formats.
* Use the same algorithm implementation for build, external print command and statusbar format strings.  This actually fixes a corner case in the statusbar implementation that would repeat part of the format if the format ended with `%`.
* Add test code for the algorithm.

---

If this is too many changes close to a release (although most of them are either legitimately useful, or smaller than they might look at first glance), I can leave out e.g. the `ui_statusbar_update()` changes that are not actually part of an original intended fix.  The only fix there is handling a trailing `%`, but it's not of huge importance, and is trivial to fix in the previous code.

But we need to fix `%p` before 2.1 at the very least.

## Hints for review

* in *ui_utils.c* (278d1901fe28bc6be7d368e4c5a6c056ec485511), check the diff ignoring whitespace changes.  A lot of it had little changes but has been unindented 1 level.
* in *utils.c*, maybe check the before and after rather than the diff itself, that looks way more confusing because of the few coincidentally identical lines, whereas the algorithm is just entirely different.

### CODE REVIEW

File: src/utils.c [hunk 1]
```diff
@@ -2499,7 +2499,24 @@ gchar *utils_get_os_info_string(void)
...
+ *                           or not.
+ * @param data User data to pass to the @p insert_replacement
+ * @returns A copy of @p str with placeholders replaced
+ *
+ * Replaces arbitrary placeholders in a printf-style format string.  Unknown
+ * placeholders are left unmodified in the resulting string.
+ * There is one special placeholder that is always handled and replaced, that
+ * is `%`: it is replaced with a literal "%"; that is `%%` results in `%`.
+ */
+GEANY_EXPORT_SYMBOL
```
Comment (techee) [inline src/utils.c:2518]: Without having a look at the rest of the patch in detail - nice! I could have used it in https://github.com/geany/geany-plugins/pull/1432 and it's nice users can expect the same beavior across Geany and plugins.

Reminder: will require API bump.

Comment (techee) [reply src/utils.c:2518]: Ah, OK, different macro. But I think it could eventually be nice as plugins (mine including) don't handle `%` escaping in any way.

**Comment by b4n** [comment]
If this there is no objections, I'd like to merge this in a couple days.  It's something I think we must fix in 2.1, so better merge it "early".  Or revert #4250 for 2.1.

**Comment by eht16** [comment]
>     * Support `%%` to insert a literal `%` in build formats.

See also https://github.com/geany/geany/pull/4250#discussion_r2030223506 about this. I'm fine with replacing `%%` as `%`.

**Comment by b4n** [comment]
> > ```
> > * Support `%%` to insert a literal `%` in build formats.
> > ```
> 
> See also [#4250 (comment)](https://github.com/geany/geany/pull/4250#discussion_r2030223506) about this. I'm fine with replacing `%%` as `%`.

Yeah the issue with the behavior you commented on was that inserting a literal `%` was not possible or unreliable, here it should be consistently handled printf-like, e.g. `%%` means `%`, however as many there might be (so `%%%%%%` means `%%%`).
So I'd think your concerns were valid, but the impl here is not subject to them.

File: src/utils.c [hunk 2]
```diff
@@ -2499,149 +2499,142 @@ gchar *utils_get_os_info_string(void)
...
 	}
-
-	if (g_utf8_strchr(needles, -1, 'l') != NULL && strstr(haystack->str, "%l")) {
-		g_string_append(errormsg, " %%l");
+	else
+	{
+		switch (placeholder)
+		{
+			case 'p': /* we only end up here if no project are open, so fallback on %d */
+				ui_set_statusbar(FALSE, _("failed to substitute %%p, no project active"));
```
Comment (eht16) [inline src/utils.c:2578]: I wonder if the message could be misleading a bit because we *did* substitute the `%p` placeholder with `%d` and it is no error.

I think we could either rephrase the message to make it more informational (e.g. `no project active, substituted with %%d`) or just drop it completely.

File: src/utils.c [hunk 3]
```diff
@@ -2499,149 +2499,142 @@ gchar *utils_get_os_info_string(void)
...
 
-	if (g_utf8_strchr(needles, -1, 'f') != NULL && strstr(haystack->str, "%f")) {
-		g_string_append(errormsg, " %%f");
+	/* %p is the project base path, if any */
+	if (placeholder == 'p' && app->project)
+		r = project_get_base_path();
+	else if (! doc || ! doc->file_name)
+	{
+		ui_set_statusbar(FALSE, _("failed to substitute %%%c: document has no path"), placeholder);
+		/* TODO: Shouldn't we rather return an empty string for known placeholders?
```
Comment (eht16) [inline src/utils.c:2570]: Returning an empty string should be ok, I think.

This case should be rather hard to reach at all.

**Review by eht16 — APPROVED** [review]
Great work.

LGTM and tested.

Thanks for fixing the placeholder mess up which I messed to notice in the review and also for even more refactoring the code to also use it for the status bar.

File: src/utils.c (hunk 2 above)
Comment (b4n) [reply src/utils.c:2578]: Good point, either sound good.

File: src/utils.c (hunk 3 above)
Comment (b4n) [reply src/utils.c:2570]: For printing it might not be hard.  For build commands it might, as we save before executing them.

Either way, the question of whether to substitute unknown placeholders or not is valid in both cases:
* when it would usually be a supported one, it's potentially a consistency question (what should happen when that valid placeholder is used in an invalid situation)
* when it would be an *unsupported* placeholder in all cases (e.g. currently `%z` for example), it has pros and cons: replacing with nothing helps noticing cases where one meant a literal `%` (which can be a problem with forward compatibility if we choose to add more placeholders in the future), but keeping the placeholder as-is can help with debugging a misspelled placeholder (and not leaving an invalid placeholder getting unnoticed as it doesn't expand at all…).

However, I don't think the previous statusbar behavior of removing the `%` but keeping the format character makes much sense.

Here I don't really have an opinion, as if we hit this it's just not gonna be very useful anyway. But if we *do*, empty is probably more "correct* I guess.

Comment (b4n) [reply src/utils.c]: I just added a change that implements replacing with nothing for known placeholders

File: src/utils.c (hunk 2 above)
Comment (b4n) [reply src/utils.c]: I updated the status message not to suggest `%p` might be invalid or not handed

File: src/utils.c (hunk 1 above)
Comment (b4n) [reply src/utils.c:2518]: Sure, if it's useful we can add it.

File: src/utils.c [hunk 4]
```diff
@@ -2499,149 +2499,150 @@ gchar *utils_get_os_info_string(void)
...
-			line_num = sci_get_current_line(doc->editor->sci) + 1;
-			replacement = g_strdup_printf("%i", line_num);
-			utils_string_replace_all(haystack, "%l", replacement);
-			g_free(replacement);
+		p++;
+		if (! *p || *p == '%') /* %% or trailing % -> % */
+			g_string_append_c(res, '%');
+		else if (! insert_replacement(res, *p, data))
+		{
+			/* unknown placeholder, leave it literally */
```
[the remark this answers is not in the document]
Comment (b4n) [reply src/utils.c:2543]: I just added this, because it seems like it'd be useful, at least for debugging

**Comment by b4n** [comment]
I updated documentation and fixed a couple minor details, and now I'd (again) say it's ready.

**Comment by eht16** [comment]
Latest changes look great, :100: for updating the docs.
